### PR TITLE
fix WSGI inclusion

### PR DESCRIPTION
**Pull request by duck-rh**
This is a proposal to implement the  changes discussed in #7

### CODE REVIEW

File: tasks/main.yml [hunk 1]
```diff
@@ -113,6 +110,13 @@
   notify: verify config and restart httpd
   when: _use_tls
 
+- name: remove obsolete files
+  file:
+    path: "{{ item }}"
+    state: absent
+  with_items:
+    - "{{ _vhost_confdir }}/python.conf"
+
```
Comment (mscherer) [inline tasks/main.yml]: Missing a notify to reload the server.

I would also like to see a comment to ay "remove this after $somedate" so we do not carry that for ever.

File: templates/vhost.conf [hunk 2]
```diff
@@ -30,4 +30,12 @@
         CustomLog logs/{{ _website_domain }}_access.log combined
 {% endif %}
 
+{% if wsgi_script is defined and wsgi_script %}
+{% if not _force_tls or port|int == 443 %}
```
Comment (mscherer) [inline templates/vhost.conf]: Why not drop that as a file in {{ _vhost_confdir }}/ ? (cause this will deal with _not_force_tls directly there) ?

**Comment by mscherer** [comment]
So, I now remember what was blocking me from fixing the PR. My understanding was that we would have wanted more than 1 wsgi script per vhost, and this change do not implement that. 

Are we ok that for later and  ok to focus on the use case of 1 vhost => 1 wsgi ? (or did I just simply misunderstood ?)

File: tasks/main.yml (hunk 1 above)
Comment (duck-rh) [reply tasks/main.yml]: My bad.

How can you possibly define $somedate to something meaningful in advance?

**Comment by duck-rh** [comment]
you're right, I just treated by use case and totally forgot about it.

I guess I can try to emulate this and allow for a mapping. I would then fail the run when using the old parameter.

I'll do this next week.

File: tasks/main.yml (hunk 1 above)
Comment (mscherer) [reply tasks/main.yml]: Well, I guess we can say we depreciate that in 6 months, 1 year ? I just like to have a expiration date to stuff we deprecate.

Realistically, all our servers will have got a run by then.

**Comment by mscherer** [comment]
For the record, I am not sure we can fix this in the httpd role. As long something drop the 3 magic directives at the right place, that should be good, and as long as there is no conflict regarding the wsgi group. 

So this can be done in the mailman role with a custom snippet The only annoyance is pulling mod_wsg manually, but I do not see a good solution right now.

**Comment by duck-rh** [comment]
So, after a new year look at this, I think we should stick to 1=>1.

To handle more in the role we would need to pass not only the webpath=>scriptpath association but also the user and other parameters, which would mean a complex structure.

What we could do is to allow the wsgi_script variable to be True, which would setup the global settings but leave the vhost-specific settings to rules outside the role. This is a minor change.

Also I would like to keep this PR as basis because it solves the inclusion whatever the force_tls velue is, and it adds documentation.

**Comment by mscherer** [comment]
So what about having a httpd-mod_wsgi role that would basically:
- requires httpd (maybe optional)
- put the global config for that in that role (ie, something that enable the wsgi config)
- drop 1 file per application doing the setup for user/alias ?

**Comment by duck-rh** [comment]
It could work yes.

So another thing which makes me uncomfortable is the management of global options. When they are calculated independently of instance settings it is fine, but we have WSGIRestrictSignal which is to be set if at least one of the instance needs it. It seems weird to repeat in the playbook all global settings even for instances which do not need it. Also in this specific case it is not per-vhost but server-global, which is even more problematic.

So I don't know how to handle the configuration on all 3 levels (instance, vhost, server) easily. Creating 3 roles seems bothersome.

In this case, because it's a boolean, we can just add the directive, if needed, along with the instance settings, and repeat it with the same value ('Off') if several instances need it. Would that be ok?

**Comment by duck-rh** [comment]
I wonder about how to share knowledge with the httpd role, like the variables containing the paths. I would rather avoid duplicating the logic and reuse the variable from the base httpd role. The problem is, if we need to do this separation for another feature and a vhost needs both, then inheritance won't work.

**Comment by mscherer** [comment]
Indeed. And making a role just for that seems cumbersome.

**Comment by duck-rh** [comment]
Our roles often do not need to do multiple tasks, but here depending on website_domain being set or not it changes behavior. Maybe we should acknowledge we need some kind of explicit variable to tell which "function" we want to call.

**Comment by mscherer** [comment]
Mhh, what do you mean "depending on website_domain", it does different things ? website_domain just have a default value of the current hostname, but it would do the same thing. I am not sure to follow you, can you explain ?

**Comment by duck-rh** [comment]
Well, this is a problem in itself as seen on #12.

I thought there was more conditional on it but there is only one:
roles/httpd/tasks/letsencrypt.yml:26:  when: website_domain is defined
which may even be a mistake but I'm not sure.

I think we try to be smart with this default vhost but in the end there is no way to setup the server and a vhost separately. Which means inheritance causes global settings to be rewritten and then make it impossible to be setup.

So in the WSGI case, if we happen to have two vhost each with its own WSGI application, then templates/python_global.conf could be rewritten in the second call with different global settings, and this is a nasty bug.

So this roles has even 3 intermixed usage:
- ensure apache is installed properly by inheriting the role
- setup global settings
- setup a specific vhost

I don't feel comfortable with this, it feels like too much magic and I keep falling into traps.

So I understand you're unwilling to change the very fundamentals of this role, with all the works and stabilization that would imply.

If we want to go on with this current role, then I think it means at least that all global parameters which could be useful to manage would have to be removed from this role and added manually in a /etc/httpd/conf.d/<app>.conf file for example. This means for WSGI that WSGIRestrictSignal should not be managed anymore.

Also either we provide a single WSGI instance per vhost, or nothing (if we don't ant to pass a complex struture of parameters). As we cannot separate the two it means even the global identical settings would have to be handled by the application. Or we could also only setup the ones in python_global.conf (except WSGIRestrictSignal) and leave instance settings to the app.

Either way it feels we're unable to provide a finished and polished WSGI helper with the current role architecture, and I'm quite disappointed (and annoyed because I can't go on with my work and this WSGI subject is really killing me for months).

**Comment by duck-rh** [comment]
See comments on #7 about upcoming solution.

**Pull request closed without merging**